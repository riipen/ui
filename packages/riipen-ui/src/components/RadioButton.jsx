import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Typography from "./Typography";

const RadioButton = ({
  checked,
  classes,
  color,
  disabled,
  label,
  id,
  prefix,
  size,
  suffix,
  ...other
}) => {
  const theme = React.useContext(ThemeContext);

  let typographyVariant;
  if (size === "small") {
    typographyVariant = "body2";
  } else if (size === "large") {
    typographyVariant = "h5";
  }

  return (
    <div
      className={clsx(
        classes,
        color,
        checked && "checked",
        disabled && "disabled",
        "root",
        size
      )}
    >
      <label
        className={clsx(checked && "checked", disabled && "disabled")}
        htmlFor={id}
      >
        <input checked={checked} disabled={disabled} type="radio" {...other} />
        <Typography
          component="span"
          variant={typographyVariant}
          color="inherit"
        >
          {prefix && <span className="content">{prefix}</span>}
          <span className="content">{label}</span>
          {suffix && <span className="content">{suffix}</span>}
        </Typography>
      </label>
      <style jsx>{`
        /* Common */

        .content {
          margin: 0 5px;
        }

        .root {
          display: inline-flex;
          cursor: pointer;
          border: 1px solid ${theme.palette.grey[300]};
          transition: 0.2s linear;
        }

        label {
          cursor: pointer;
          text-align: center;
        }

        input {
          opacity: 0;
          position: absolute;
          pointer-events: none;
        }

        .disabled {
          cursor: default;
        }

        /* Sizes */

        .small label {
          min-width: 100px;
          padding: 4px;
          width: 100px;
        }

        .medium label {
          min-width: 150px;
          padding: 8px;
          width: 150px;
        }

        .large label {
          min-width: 200px;
          padding: 10px;
          width: 200px;
        }

        /* Colors */

        .root {
          background-color: transparent;
          border-color: ${theme.palette.grey[300]};
          color: ${theme.palette.text.secondary};
        }

        .root.default:hover {
          border-color: ${theme.palette.grey[600]};
          color: ${theme.palette.text.primary};
        }

        .root.default.checked {
          background-color: ${theme.palette.grey[300]};
          border-color: ${theme.palette.grey[300]};
          color: ${theme.palette.common.black};
        }

        .root.primary:hover {
          border-color: ${theme.palette.primary.light};
          color: ${theme.palette.primary.dark};
        }

        .root.primary.checked {
          background-color: ${theme.palette.primary.main};
          border-color: ${theme.palette.primary.main};
          color: ${theme.palette.common.white};
        }

        .root.secondary:hover {
          border-color: ${theme.palette.secondary.light};
          color: ${theme.palette.secondary.dark};
        }

        .root.secondary.checked {
          background-color: ${theme.palette.secondary.main};
          border-color: ${theme.palette.secondary.main};
          color: ${theme.palette.common.white};
        }

        .root.tertiary:hover {
          border-color: ${theme.palette.tertiary.light};
          color: ${theme.palette.tertiary.dark};
        }

        .root.tertiary.checked {
          background-color: ${theme.palette.tertiary.main};
          border-color: ${theme.palette.tertiary.main};
          color: ${theme.palette.common.white};
        }

        .root.disabled:hover,
        .root.disabled {
          background-color: ${theme.palette.disabled};
          border-color: ${theme.palette.disabled};
          color: ${theme.palette.grey[200]};
        }
      `}</style>
    </div>
  );
};

RadioButton.propTypes = {
  /**
   * If `true`, the component is checked.
   */
  checked: PropTypes.bool,

  /**
   * An array of custom CSS classes to apply.
   */
  classes: PropTypes.array,

  /**
   * The color of the radio button.
   */
  color: PropTypes.oneOf(["default", "primary", "secondary", "tertiary"]),

  /**
   * If `true`, the radio will be disabled.
   */
  disabled: PropTypes.bool,

  /**
   * The id of the input component.
   */
  id: PropTypes.string,

  /**
   * Label text to display for the radio.
   */
  label: PropTypes.node,

  /**
   * The component to render before the label text.
   */
  prefix: PropTypes.node,

  /**
   * The size of the radio button
   */
  size: PropTypes.oneOf(["small", "medium", "large"]),

  /**
   * The component to render after the label text.
   */
  suffix: PropTypes.node
};

RadioButton.defaultProps = {
  size: "medium",
  color: "default"
};

RadioButton.displayName = "RadioButton";

export default withClasses(RadioButton);
