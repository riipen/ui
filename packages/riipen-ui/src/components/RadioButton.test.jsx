import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import { faFlag } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import RadioButton from "./RadioButton";

const icon = i => <FontAwesomeIcon icon={i} />;

describe("<RadioButton>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<RadioButton />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<RadioButton />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("checked prop", () => {
    it("sets element's class name or prop value when checked is true", () => {
      const checked = true;

      const wrapper = mount(<RadioButton checked={checked} readOnly />);

      expect(
        wrapper
          .find("RadioButton")
          .childAt(0)
          .hasClass("checked")
      ).toEqual(true);
      expect(wrapper.find("label").hasClass("checked")).toEqual(true);
      expect(wrapper.find("input").props().checked).toEqual(checked);
      expect(wrapper.find("Typography").props().color).toEqual("initial");
    });

    it("sets color as grey600 when checked is false", () => {
      const checked = false;

      const wrapper = mount(<RadioButton checked={checked} readOnly />);

      expect(wrapper.find("Typography").props().color).toEqual("grey600");
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<RadioButton classes={classes} />);

      expect(
        wrapper
          .find("RadioButton")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-radiobutton"];

      const wrapper = mount(<RadioButton />);

      expect(
        wrapper
          .find("RadioButton")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("disabled prop", () => {
    it("sets element's class name or disabled value when disabled is true", () => {
      const disabled = true;

      const wrapper = mount(<RadioButton disabled={disabled} readOnly />);

      expect(
        wrapper
          .find("RadioButton")
          .childAt(0)
          .hasClass("disabled")
      ).toEqual(true);
      expect(wrapper.find("label").hasClass("disabled")).toEqual(true);
      expect(wrapper.find("input").props().disabled).toEqual(disabled);
    });
  });

  describe("id prop", () => {
    it("sets custom id", () => {
      const id = "Test";

      const wrapper = mount(<RadioButton id={id} readOnly />);

      expect(wrapper.find("input").props().id).toEqual(id);
      expect(wrapper.find("label").props().htmlFor).toEqual(id);
    });
  });

  describe("label prop", () => {
    it("sets custom label", () => {
      const label = "Test";

      const wrapper = mount(<RadioButton label={label} readOnly />);

      expect(wrapper.text()).toEqual(label);
      expect(
        wrapper
          .find(".content")
          .children()
          .contains(label)
      ).toEqual(true);
    });
  });

  describe("prefix prop", () => {
    it("sets custom prefix", () => {
      const prefix = icon(faFlag);

      const wrapper = mount(<RadioButton prefix={prefix} readOnly />);

      expect(
        wrapper
          .find(".content")
          .children()
          .contains(prefix)
      ).toEqual(true);
    });
  });

  describe("suffix prop", () => {
    it("sets custom suffix", () => {
      const suffix = icon(faFlag);

      const wrapper = mount(<RadioButton suffix={suffix} readOnly />);

      expect(
        wrapper
          .find(".content")
          .children()
          .contains(suffix)
      ).toEqual(true);
    });
  });
});
