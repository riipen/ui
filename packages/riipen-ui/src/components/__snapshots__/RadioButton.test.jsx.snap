// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton> renders correct snapshot 1`] = `
<withClasses(RadioButton)>
  <RadioButton
    classes={
      Array [
        "riipen",
        "riipen-radiobutton",
      ]
    }
  >
    <div
      className="jsx-1306841623 riipen riipen-radiobutton root"
    >
      <label
        className="jsx-1306841623 "
      >
        <input
          className="jsx-1306841623 "
          type="radio"
        />
        <withClasses(Typography)
          color="grey600"
          variant="body2"
        >
          <Typography
            classes={
              Array [
                "riipen",
                "riipen-typography",
              ]
            }
            color="grey600"
            display="initial"
            gutter={false}
            textAlign="inherit"
            textTransform="inherit"
            variant="body2"
          >
            <p
              className="jsx-42457783 root color-grey600 initial body2 align-inherit transform-inherit riipen riipen-typography"
            >
              <span
                className="jsx-1306841623 content"
              />
            </p>
            <JSXStyle
              dynamic={
                Array [
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "14px",
                  "20px",
                  "0.00938em",
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "12px",
                  "17px",
                  "0.01071em",
                  "#424242",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "10px",
                  "14px",
                  "0.01071em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "40px",
                  "56px",
                  "-0.01562em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "30px",
                  "42px",
                  "-0.00833em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  500,
                  "22px",
                  "32px",
                  "0em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "20px",
                  "30px",
                  "-0.00735em",
                  "#373737",
                  "Roboto, Helvetica, Arial, sans-serif",
                  400,
                  "16px",
                  "23px",
                  "0rem",
                  "#000",
                  "#cccbcb",
                  "#747474",
                  "#424242",
                  "#373737",
                  "#1f71ae",
                  "#f58632",
                  "#3caabb",
                  "#4cc88d",
                  "#e95353",
                  "#febb46",
                  "#fff",
                  700,
                  300,
                  500,
                  400,
                ]
              }
              id="1840119567"
            />
          </Typography>
        </withClasses(Typography)>
      </label>
      <JSXStyle
        dynamic={
          Array [
            "#dddddd",
            "#eeeeee",
          ]
        }
        id="547299733"
      />
    </div>
  </RadioButton>
</withClasses(RadioButton)>
`;
